### PR TITLE
Allowing Vive controllers and trackers to be used in desktop mode

### PR DESCRIPTION
**Pull request by druiz17**
Allowing Vive controllers and trackers to be used in desktop mode

### CODE REVIEW

**Comment by CainFoool** [comment]
Testing when ready.

**Comment by druiz17** [comment]
@CainFoool not ready for testing

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1258/HighFidelity-Beta-PR10866-6f30cb18256bec02cf70193097d92a93b47f6c57.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1258/HighFidelity-Beta-PR10866-6f30cb18256bec02cf70193097d92a93b47f6c57.exe

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1264/HighFidelity-Beta-PR10866-6f30cb18256bec02cf70193097d92a93b47f6c57.dmg

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1266/HighFidelity-Beta-PR10866-6f30cb18256bec02cf70193097d92a93b47f6c57.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1266/HighFidelity-Beta-PR10866-6f30cb18256bec02cf70193097d92a93b47f6c57.exe

**Comment by druiz17** [comment]
Test plan
- this PR is only testing the head tracker in desktop mode. Any other configuration does not work in this PR and will be fixed in another PR
- goto to the calibration ui
- select the head tracker and the option "Use vive devices in desktop mode"
- then perform the calibration.
- you should be able to move your head and hand in desktop mode and have room scale capabilities

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 1]
```diff
@@ -218,6 +248,18 @@ void ViveControllerManager::pluginUpdate(float deltaTime, const controller::Inpu
         return;
     }
 
+    if (isDesktopMode() && _desktopMode) {
+        if (!_resetMatCalculated) {
+            _resetMat = calculateResetMat();
+            _resetMatCalculated = true;
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: So, I never see where you set this to false.  Will there be a problem if the user switches from desktop to HMD and back?

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 2]
```diff
@@ -718,10 +760,6 @@ glm::mat4 ViveControllerManager::InputDevice::calculateDefaultToReferenceForHead
 
     glm::mat4 finalHeadPuck = newHeadPuck * headPuckOffset;
 
-    glm::mat4 defaultHeadOffset = glm::inverse(inputCalibration.defaultCenterEyeMat) * inputCalibration.defaultHeadMat;
-
-    glm::mat4 currentHead  = finalHeadPuck * defaultHeadOffset;
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: currentHead has been removed, does this even compile?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1331/HighFidelity-Beta-PR10866-27a9880eff2fc7d9b4e24578ab0794bfddc8f033.dmg

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 3]
```diff
@@ -761,7 +764,7 @@ glm::mat4 ViveControllerManager::InputDevice::calculateDefaultToReferenceForHead
     glm::mat4 finalHeadPuck = newHeadPuck * headPuckOffset;
 
     // calculate the defaultToRefrenceXform
-    glm::mat4 defaultToReferenceMat = currentHead * glm::inverse(inputCalibration.defaultHeadMat);
+    glm::mat4 defaultToReferenceMat = finalHeadPuck * glm::inverse(inputCalibration.defaultHeadMat);
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: Hey, are you sure about this,  I could be wrong, but I thought the first offset was between the puck and the center-eye, and the second offset was from the center-eye to the avatar's head.  IMHO, I think you need both to handle avatars with very large heads.

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 4]
```diff
@@ -60,10 +62,10 @@ static const int SECOND_FOOT = 1;
...
 static const int CHEST = 3;
 
-static float HEAD_PUCK_Y_OFFSET = -0.0254f;
-static float HEAD_PUCK_Z_OFFSET = -0.152f;
-static float HAND_PUCK_Y_OFFSET = -0.0508f;
-static float HAND_PUCK_Z_OFFSET = 0.0254f;
+static float HEAD_PUCK_Y_OFFSET = 0.0f;
+static float HEAD_PUCK_Z_OFFSET = 0.0f;
+static float HAND_PUCK_Y_OFFSET = 0.0f;
+static float HAND_PUCK_Z_OFFSET = 0.0f;
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: These offsets are not constant, but are uppercase.  This confused me.  They probably should be member variables.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1340/HighFidelity-Beta-PR10866-07d098b95beb28d4b3993884113f47158fee785f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1340/HighFidelity-Beta-PR10866-07d098b95beb28d4b3993884113f47158fee785f.dmg

**Comment by anshuman64** [comment]
QA Approved with known exceptions:
1. Head puck with hand controllers only works right now.
2. Rotating head beyond 90 degrees in 3rd person camera mode causes the Avatar IK fail.
Tony will take this across the finish line to work with all pucks. I'm creating a low-priority ticket for #2 and we will address eventually.

**Pull request merged**